### PR TITLE
pageserver: fix vectored read aux key handling

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem
Vectored get would descend into ancestor timelines for aux files.
This is not the behaviour of the legacy read path and blocks cutting
over to the vectored read path.

Fixes https://github.com/neondatabase/neon/issues/7379

## Summary of Changes
Treat non inherited keys specially in vectored get. At the point when
we want to descend into the ancestor mark all pending non inherited keys
as errored out at the key level. Note that this diverges from the
standard vectored get behaviour for missing keys which is a top level
error. This divergence is required to avoid blocking compaction in case
such an error is encountered when compaction aux files keys. I'm pretty
sure the bug I just described predates the vectored get implementation,
but it's still worth fixing.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [inline pageserver/src/tenant/timeline.rs:3870]: need fix after split missing key pull request gets merged

**Review by skyzh — COMMENTED** [review]
my main concern of this pull request is that we also have a `is_inherited_key` function, which serves almost the same purpose of `NON_INHERITED_RANGE`. I'm wondering if there could be some hints like "if you change `is_inherited_key`, also change `NON_INHERITED_RANGE`", or have a single place to define both things, so as to avoid problems in the future.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:875]: Indeed. I wrote this PR such that it's not blocked by https://github.com/neondatabase/neon/pull/7393, but I'm ok with waiting for that to merge.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:2975]: > NON_INHERITED_RANGE will be a keyspace containing multiple ranges in the future

Hmm. I assumed it would be a single range. Regardless, the same approach should work. The `overlaps` check on line 2975 is just an optimisation (`remove_overlapping_with` can be expensive). The check could simply be removed.

**Comment by VladLazar** [comment]
> my main concern of this pull request is that we also have a `is_inherited_key` function, which serves almost the same purpose of `NON_INHERITED_RANGE`. I'm wondering if there could be some hints like "if you change `is_inherited_key`, also change `NON_INHERITED_RANGE`", or have a single place to define both things, so as to avoid problems in the future.

Yeah that's true. Let's define `is_inherited_key` in terms of `NON_INHERITED_RANGE`: `!NON_INHERITED_RANGE.contains(&key)`.

File: pageserver/src/tenant/timeline.rs
Comment (skyzh) [reply pageserver/src/tenant/timeline.rs:2975]: as long as we fully disable aux v1 it will become a single range

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:3937]: Fixed in https://github.com/neondatabase/neon/pull/7404/commits/020153a376731e46d3a1b7ddd6029045d9018873

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:3038]: I'll leave this change up to you since I'm not sure what you'll need for aux v2.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: https://github.com/neondatabase/neon/pull/7404/commits/020153a376731e46d3a1b7ddd6029045d9018873

File: libs/pageserver_api/src/key.rs
Comment (skyzh) [inline libs/pageserver_api/src/key.rs:485]: I assume this would be less efficient than doing a != check but I'm fine with that...

**Review by skyzh — APPROVED** [review]
Rest LGTM

File: libs/pageserver_api/src/key.rs
Comment (VladLazar) [reply libs/pageserver_api/src/key.rs:485]: Yeah, it's an extra comparison in the worst case, but I wouldn't be surprised if the compiler figures out that: `x <= y < x + 1` is equivalent to `x == y`. I wrote a compiler optimisation which did something similar in the past.

File: pageserver/src/tenant/timeline.rs
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: [This](https://github.com/neondatabase/neon/blob/e69ff3fc00ab8be31e8f69eb3726da1b83d84180/pageserver/src/pgdatadir_mapping.rs#L803-L805) check in collect keyspace ensures that compaction will be able to read the aux file key. Hence the check is not needed and was removed.

File: pageserver/src/tenant/timeline.rs
Comment (arpad-m) [inline pageserver/src/tenant/timeline.rs:3046]: I'm wondering about this part. Right now, `NON_INHERITED_RANGE` is just one key so this loop is trivial. However, with aux file storage v2 that range would become quite large (120 bits I guess?) and also be nonconvex (have gaps), so if someone tried reading the aux file range, this loop over the entire not found range might turn into a footgun.

**Review by arpad-m — APPROVED** [review]
We can merge this in my opinion, but I don't think it being a range helps us a lot: we'll need support for a list of ranges in the future instead of just one. That being said, refactoring from this state should not be harder than refactoring from a simpler version of this PR that bakes in the assumption that `NON_INHERITED_RANGE` is just one element large.

File: pageserver/src/tenant/timeline.rs
Comment (arpad-m) [reply pageserver/src/tenant/timeline.rs:3046]: Thanks, that makes sense.

**Pull request merged**
